text/3727-unsafe-set-discriminant.md

```diff
@@ -0,0 +1,320 @@
+
+However, when authoring generic or macro code, which may potentially accept types that do not have these qualities, it is necessary to call `set_discriminant` to fully initialize the enum in the general case.
+
+For example, if a macro was used to populate an `Option<U>` value, and users could chose `U: u32` (which does not have a niche repr), OR `U: &u32` (which does have a niche repr), then the author of the macro should call `set_discriminant` to soundly initialize the `Option<U>` in all cases.
+
+[primitive representation]: https://doc.rust-lang.org/reference/items/enumerations.html#pointer-casting
+[discriminant elision]: https://rust-lang.github.io/unsafe-code-guidelines/layout/enums.html#discriminant-elision-on-option-like-enums
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* This RFC increases the API surface of the standard library
+* This RFC adds unsafe methods which must be considered, documented, and tested
+* The abilities of this RFC could technically already be done today, either for carefully selected subsets of enums (specific reprs, hand-built enums using explicit discriminant and `union`'d value fields), or be done today less efficiently with explicit value creations and copies
+* The features added to this RFC generally only benefit "power users", likely library authors already doing unsafe things
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This RFC is necessary to "complete" the ability to soundly create enums in-place.
+
+Without this capability, downstream library authors are required to make a choice between:
+
+* Suboptimal stack usage and performance, in the case of by-value initialization
+* Restriction of functionality to a subset of enum reprs, or bespoke proxy types that emulate enums to allow for in-place construction
+
+This is necessary to be part of the language and/or standard library as enum representations (at least for the default `repr(Rust)`) are unspecified, and particularly with concepts like niche-packed enums, it is not possible to soundly handle this in the general case of user defined enum types.
+
+I am not aware of any other proposed designs for this functionality.
+
+# Prior art
+[prior-art]: #prior-art
+
+The `rkyv` crate defines [proxy enums] with explict `#[repr(N)]` (where N is an unsigned integer) types, in order to allow in-place construction and access
+
+[proxy enums]: https://rkyv.org/format.html
+
+Syntax for `discriminant_of!` is inspired by the in-progress [offset_of_enum feature][feature_offset_of_enum].
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Should we provide alternate forms of `set_discriminant`?
+
+There was some discussion when writing this RFC what the type of the first argument to `set_discriminant` should be:
+
+* `&mut MaybeUninit<T>`
+* `NonNull<T>`
+* `*mut T`
+
+`*mut T` was chosen as the most general option, however it is likely desirable to accept other forms as well for convenience.
```

Comment:
In the original draft, I had proposed using `&mut MaybeUninit<T>`, rather than `*mut T`. When I swapped that over I left this section, but refactored it. I probably could remove it, but there is still some open discussion about what API should there be on day one.

I'll likely remove this in the near future, as I am somewhat convinced that if there were to be only one interface, it should be on `*mut T`.